File: JavaScript/PathSumIII/PathSumIII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
// (168ms)
var pathSum = function(root, sum) {
    if(root === null)
        return 0;
    return helper(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
};

var helper = function(root, sum) {
        if(root === null)
            return 0;
        var res = 0;
        if(root.val === sum)
            res++;
        res += helper(root.left, sum - root.val) + helper(root.right, sum - root.val);
        return res;
};

// (232ms)
// 可见javascript的map效率多慢
var pathSum2 = function(root, sum){
    var preSum = new Map();
    preSum.set(0, 1);
    return helper2(root, 0, sum, preSum);
};

var helper2 = function(root, curSum, sum, preSum) {
    if(root === null)
        return 0;
    curSum += root.val;
    var res = 0;
    if(preSum.has(curSum - sum))
        res += preSum.get(curSum - sum);
    if(preSum.has(curSum)) {
        preSum.set(curSum, preSum.get(curSum) + 1);
    } else {
        preSum.set(curSum, 1);
    }
    res += helper2(root.left, curSum, sum, preSum);
    res += helper2(root.right, curSum, sum, preSum);
    preSum.set(curSum, preSum.get(curSum) - 1);
    return res;
};

File: Python/PathSumIII/PathSumIII.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    # (1428ms)
    # this solution is much slower
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if not root:
            return 0
        return self.helper(root, sum) + self.pathSum(root.left, sum) + self.pathSum(root.right, sum)

    def helper(self, nowNode, nowSum):
        if not nowNode:
            return 0
        res = 1 if nowNode.val == nowSum else 0
        res += self.helper(nowNode.left, nowSum - nowNode.val) \
            + self.helper(nowNode.right, nowSum - nowNode.val)
        return res

    # (126ms)
    # this solution is much better
    def pathSum2(self, root, sum):
        preSum = {0: 1}
        return self.helper2(root, 0, sum, preSum)

    def helper2(self, root, curSum, target, preSum):
        if not root:
            return 0
        curSum += root.val
        res = preSum.get(curSum - target, 0)
        preSum[curSum] = preSum.get(curSum, 0) + 1
        res += self.helper2(root.left, curSum, target, preSum)
        res += self.helper2(root.right, curSum, target, preSum)
        preSum[curSum] -= 1
        return res
